### ISSUE TITLE
[Netfx]: Bug in AsyncStreamReader.GetLinesFromStringBuilder() private API

### ISSUE THREAD

**Issue by safern** [dotnet/corefx#18437]
When running the following code, we would get an `IndexOutOfRangeException` from `StringBuilder.get_Char(char)`.

```cs
        [Fact]
        public void TestAsyncHalfCharacterAtATime()
        {
            var receivedOutput = false;
            var collectedExceptions = new List<Exception>();

            Process p = CreateProcess(() =>
            {
                var stdout = Console.OpenStandardOutput();
                var bytes = new byte[] { 97, 0 }; //Encoding.Unicode.GetBytes("a");

                for (int i = 0; i != bytes.Length; ++i)
                {
                    stdout.WriteByte(bytes[i]);
                    stdout.Flush();
                    Thread.Sleep(100);
                }
                return SuccessExitCode;
            });
            p.StartInfo.RedirectStandardOutput = true;
            p.StartInfo.StandardOutputEncoding = Encoding.Unicode;
            p.OutputDataReceived += (s, e) =>
            {
                try
                {
                    if (!receivedOutput)
                    {
                        Assert.Equal(e.Data, "a");
                    }
                    receivedOutput = true;
                }
                catch (Exception ex)
                {
                    // This ensures that the exception in event handlers does not break
                    // the whole unittest
                    collectedExceptions.Add(ex);
                }
            };
            p.Start();
            p.BeginOutputReadLine();

            Assert.True(p.WaitForExit(WaitInMS));
            p.WaitForExit(); // This ensures async event handlers are finished processing.

            Assert.True(receivedOutput);
        }
```

When we call `p.BeginOutputReadLine()` it will call `AsyncStreamReader.ReadBuffer(IAsyncResult)` everything at this moment behaves as expected which has 1 char to read in the byte buffer (97). But when reading the chars through the decoder:

https://referencesource.microsoft.com/#System/services/monitoring/system/diagnosticts/AsyncStreamReader.cs,192

We hit a bug in `UnicodeDecoder.GetChars` (see #18436) so we would get 0 as a result when reading the chars out of the decoder. After trying to decode the chars we call `GetLinesFromStringBuilder()` when in this situation where `sb.Length == 0` and `currentIndex == 0` we will try to access `sb[currentIndex - 1]` which will obviously will be an `IndexOutOfRangeException`... source code:

https://referencesource.microsoft.com/#System/services/monitoring/system/diagnosticts/AsyncStreamReader.cs,242

In .NET Core this was fixed by doing a check if currentIndex > 0 before trying to access the StringBuilder.

**Comment by safern** [comment]
Closing as netfx-port-consider label opens it on TFS

**Comment by danmosemsft** [comment]
Looks like the fix mentioned above was 17c3909e51c762881094d2cd52bdfd7c9df21d3b
